### PR TITLE
Reenable the memory eviction code

### PR DESCRIPTION
**Pull request by bosilca**
While adding support for coroutines and their use of _AGAIN to resubmit the same task onto the same stream, we have broken the support for using _AGAIN to trigger the memory evictions when the GPU memory was completely full.

This PR adds the use of _NEXT as a legitimate return code from the stream's progress_fct to indicate that the task should be kept into the stream queues but the error should be escalated upstream. Upon such an error the higher-level scheduler should trigger the memory eviction procedure asap

### CODE REVIEW

**Comment by bosilca** [comment]
We need to have a discussion about the right approach here, especially with regard to the task that triggered the need for the memory cleanup. Right now this tasks remains in the stream queue and will therefore be rescheduled for execution relatively soon (as the stream pending queue is relatively small). This will lead to a lot of noise when the GPU is trying to release memory.

We could instead also evict the task and let the upper-level scheduler handle the case. This will give us a little bit more time to find memory for a task that we know was missing some, but could also change the way tasks are ordered for execution (in the case where we hit memory constraints).

File: parsec/mca/device/device_gpu.c [hunk 1]
```diff
@@ -1884,13 +1883,24 @@ parsec_device_progress_stream( parsec_device_gpu_module_t* gpu_device,
     rc = progress_fct( gpu_device, task, stream );
     if( 0 > rc ) {
         if( PARSEC_HOOK_RETURN_AGAIN != rc ) {
-           *out_task = task;
+            if( PARSEC_HOOK_RETURN_NEXT == rc ) {
+                /* Dont reorder the push_back, we are running into physical contraints and need to delay
+                 * the resubmission of this task as much as possible, but without loosing track of it
+                 * (aka. returning it to the upper level).
+                 */
+                parsec_list_push_back(stream->fifo_pending, (parsec_list_item_t*)task);
```
[the remark this answers is not in the document]
Comment (bosilca) [reply parsec/mca/device/device_gpu.c:1891]: The comment above tries to explain the reason for not ordering. OK, assuming the reader knows that the PUSH_BACK macro could do an ordered push while the push_back just shove it at the end of the FIFO.

[the remark this answers is not in the document]
Comment (bosilca) [reply parsec/mca/device/device_gpu.c:1891]: Not sure how the fifo became malformed we do call `PARSEC_LIST_ITEM_SINGLETON` when we extract the task from the queue and as you noted the different flavors of `push_back` are both correctly linking the double-linked list.

**Comment by devreal** [comment]
Can this be merged? The discussion can be had later (if someone pushes it) but right now this is broken and a blocker for MRA in TTG.

**Review by abouteiller — APPROVED** [review]

**Pull request merged**